### PR TITLE
fix: improve client upload speed

### PR DESCRIPTION
**Pull request by maqi**
### Summary generated by Reviewpad on 14 Jun 23 16:22 UTC
This pull request contains client improvements including reduced memory usage during file uploading and improved upload speed. Changes were made in `sn_client/src/api.rs`, `sn_client/src/file_apis.rs`, `sn_networking/src/cmd.rs`, and `sn_networking/src/lib.rs`. In `sn_client/src/file_apis.rs`, the maximum batch size for chunk uploads was increased to 64, and uploading progress is now logged. In `sn_networking/src/lib.rs`, a new function was added to send requests to half of the closest peers.

### CODE REVIEW

File: sn_client/src/api.rs [hunk 1]
```diff
@@ -302,10 +314,18 @@ impl Client {
         Ok(responses)
     }
 
-    pub(crate) async fn send_and_wait_till_first_rsp(&self, request: Request) -> Result<Response> {
+    pub(crate) async fn send_half_range_and_wait_till_first_rsp(
```
Comment (joshuef) [inline sn_client/src/api.rs]: Can we get some docs on this and why we'd use it?

File: sn_client/src/file_apis.rs [hunk 2]
```diff
@@ -25,7 +25,7 @@ use tracing::trace;
 use xor_name::XorName;
 
 // Maximum number of concurrent chunks to be uploaded/retrieved for a file
-const CHUNKS_BATCH_MAX_SIZE: usize = 5;
+const CHUNKS_BATCH_MAX_SIZE: usize = 16;
```
Comment (joshuef) [inline sn_client/src/file_apis.rs]: Can we get a bit of data on why this was chosen?

File: sn_client/src/file_apis.rs [hunk 3]
```diff
@@ -170,8 +170,10 @@ impl Files {
         let payment = payment_proofs.get(&address.name().0).cloned();
         // TODO: re-enable requirement to always provide payment proof
         //.ok_or(super::Error::MissingPaymentProof(address))?;
-
-        self.client.store_chunk(chunk, payment).await?;
+        let closest_peers = self.client.client_get_closest_peers_from_local().await?;
```
Comment (joshuef) [inline sn_client/src/file_apis.rs]: If we do this all the time, when/how do we populate enough info from the network? 

What happens if this data is stale?

File: sn_networking/src/lib.rs [hunk 4]
```diff
@@ -520,6 +520,21 @@ impl Network {
             .await)
     }
 
+    /// Send `Request` to half of the closest peers. `Self` is not present among the recipients.
```
Comment (joshuef) [inline sn_networking/src/lib.rs]: Can we get docs on when/why

File: sn_networking/src/lib.rs [hunk 5]
```diff
@@ -627,6 +642,27 @@ impl Network {
         sort_peers_by_address(closest_peers, key, CLOSE_GROUP_SIZE)
     }
 
+    /// Based on the local knowledge, returns half closest peers to the given `XorName`,
+    /// sorted by their distance to the xor_name.
+    /// If `client` is false, then include `self` among the `closest_peers`
+    pub async fn client_get_closest_peers_from_local(
```
Comment (joshuef) [inline sn_networking/src/lib.rs]: this naming should probably reflect the `half`

**Review by joshuef — CHANGES_REQUESTED** [review]

File: sn_client/src/file_apis.rs (hunk 3 above)
Comment (maqi) [reply sn_client/src/file_apis.rs]: > If we do this all the time, when/how do we populate enough info from the network?

Here only the result from the `libp2p::get_closest_peers` got trimmed for the half-range result.
And it is only `libp2p::get_closest_peers` will populate network knowledge.
So no affect to the network info pupulation.

> What happens if this data is stale?

The nodes received the `Cmd::StoreChunk` shall carry out a `libp2p::put_record`, which will broadcast the copy to all 8 closest.
So, the data won't stale? or, have the simiar chance of stale as current?

File: sn_client/src/file_apis.rs (hunk 2 above)
Comment (maqi) [reply sn_client/src/file_apis.rs]: following is some comparision result from benchmark:
```
                            max_batch_size                 1MB throughput                  10MB throughput
main branch                     5                                0.57MB/s                         3.4MB/s
    PR 390                      5                                0.88MB/s                         5.8MB/s
    PR 390                     16                               0.66MB/s                         6.3MB/s
    PR 390                     32                               0.7MB/s                           6.8MB/s
```
hence set this value to 32 currently.

**Comment by maqi** [comment]
The increased `CLIENT_AVG_MEM_LIMIT_MB` is just because of the improved speed, there is less mem_usage check points to spread out the peak mem usage at the beginning due to the self_encryption

**Review by joshuef — APPROVED** [review]

**Pull request merged**